### PR TITLE
Refactor mixed mode methods

### PR DESCRIPTION
**Pull request by davidyell**
References #51 

This pull request aims to address the software design issue in the plugin of mixed mode methods. These methods allow both setting and getting of class properties. Since CakePHP 3.4+ it's better to follow the core and implement explicit setter and getter methods to allow for a more verbose, obvious and easier to read public api.

I have tried to implement the methods as I understood them. However in some instances there are both mixed input and mixed output methods, which makes refactoring difficult.

I'm unsure if mixed input and mixed output would constitute two new methods, so in those cases I have tried to follow the existing api as much as possible, to ensure the expected return types will be what existing code is expecting.

I have made one public api change in this pull request, which is changing `Endpoint::endpoint()` to `Endpoint::get/setName()` which better confers both the implied functionality and the types the method deals with.

### CODE REVIEW

File: src/AbstractDriver.php [hunk 1]
```diff
@@ -144,27 +187,55 @@ public function logQueries($enable = null)
         $this->_logQueries = $enable;
     }
 
+    /**
+     * Enable query logging for the driver
+     *
+     * @return void
```
Comment (ADmad) [inline src/AbstractDriver.php]: Should return `$this` instead of void to allow method chaining.

File: src/AbstractDriver.php [hunk 2]
```diff
@@ -144,27 +187,55 @@ public function logQueries($enable = null)
...
+     */
+    public function enableQueryLogging()
+    {
+        $this->_logQueries = true;
+    }
+
+    /**
+     * Disable query logging for the driver
+     *
+     * @return void
```
Comment (ADmad) [inline src/AbstractDriver.php]: Should return `$this` instead of void to allow method chaining.

File: src/Model/Endpoint.php [hunk 3]
```diff
@@ -133,35 +133,35 @@ class Endpoint implements RepositoryInterface, EventListenerInterface, EventDisp
...
         if (!empty($config['displayField'])) {
-            $this->displayField($config['displayField']);
+            $this->setDisplayField($config['displayField']);
         }
-        if (!empty($config['endpoint'])) {
-            $this->endpoint($config['endpoint']);
+        if (!empty($config['inflect'])) {
+            $this->setInflectionMethod($config['inflect']);
+        }
+        if (!empty($config['name'])) {
```
Comment (ADmad) [inline src/Model/Endpoint.php]: To maintain BC you should also check for old name `endpoint` and if provided set `name` using that.

File: src/Model/Endpoint.php [hunk 4]
```diff
@@ -451,66 +635,103 @@ public function resourceClass($name = null)
...
+    public function setWebservice($webservice)
+    {
+        $connection = $this->getConnection();
+        if (!$connection) {
+            throw new UnexpectedDriverException(__('No driver has been defined for this endpoint'));
+        }
+
+        $this->_webservice = $connection->webservice($webservice);
+
+        return $this->getWebservice();
```
Comment (ADmad) [inline src/Model/Endpoint.php]: Should return `$this` instead.

File: src/Model/Endpoint.php [hunk 5]
```diff
@@ -1192,15 +1405,37 @@ public function buildRules(RulesChecker $rules)
...
+            'connectionName' => $conn ? $conn->configName() : null,
+            'inflector' => $this->getInflectionMethod()
         ];
     }
+
+    /**
+     * Set the endpoint elias
+     *
+     * @param string $alias Alias for this endpoint
+     * @return void
```
Comment (ADmad) [inline src/Model/Endpoint.php]: Should return `$this`.

File: tests/TestCase/Model/EndpointTest.php [hunk 6]
```diff
@@ -40,13 +40,25 @@ public function setUp()
...
-        $this->assertSame('user-groups', $endpoint->endpoint());
+    /**
+     * @dataProvider providerEndpointNames
+     * @param string $name
+     * @param string|null $inflector
+     * @param string $expected
+     */
+    public function testEndpointName($name, $inflector, $expected)
+    {
+        $endpoint = new Endpoint(['name' => $name, 'inflect' => $inflector]);
```
Comment (ADmad) [inline tests/TestCase/Model/EndpointTest.php]: Add test using old `endpoint` key too to ensure BC.

**Comment by ADmad** [comment]
You'll need to rebase the branch on top of upstream/dev and fix merge conflicts.

File: src/Model/Endpoint.php (hunk 4 above)
Comment (davidyell) [reply src/Model/Endpoint.php]: I'm not sure that it can. [Due to this](https://github.com/davidyell/Webservice/blob/cake-version-upgrade/src/Model/Endpoint.php#L716). As the Endpoint sets the Webservice by calling the Connection, and the Connection proxies the call through to the driver. It would mean that `Connection::setWebservice()` would be returning an instance of the Driver and not an instance of the Webservice.

**Comment by davidyell** [comment]
The PHPCS failures are resolved in #56

File: src/AbstractDriver.php [hunk 7]
```diff
@@ -117,13 +117,13 @@ public function webservice($name, WebserviceInterface $webservice = null)
...
      * @param \Muffin\Webservice\Webservice\WebserviceInterface $webservice Instance of the webservice
-     * @return $this
+     * @return \Muffin\Webservice\Webservice\WebserviceInterface
      */
     public function setWebservice($name, WebserviceInterface $webservice)
     {
         $this->_webservices[$name] = $webservice;
 
-        return $this;
+        return $this->_webservices[$name];
```
Comment (ADmad) [inline src/AbstractDriver.php]: Why did you change this? Returning the webservice instance which was passed as argument is not very useful. All setter methods should return `$this` to maintain consistency and intuitiveness

Comment (davidyell) [reply src/AbstractDriver.php]: I got confused by myself because I had made a mistake when refactoring.

[This line](https://github.com/davidyell/Webservice/blob/refactor/mixed-mode-methods/src/Model/Endpoint.php#L733) was using a call to `setWebservice()` via the Connection, and the Connection proxies to the Driver. So I'd updated it to preserve the return.

However, when I double checked my work, I realised that I should have used a getter, and obviously forgot to update this.

File: src/AbstractDriver.php [hunk 8]
```diff
@@ -60,36 +60,80 @@ public function __construct($config = [])
...
+     * Set the client instance this driver will use to make requests
+     *
+     * @param object $client Client instance
+     * @return object
+     */
+    public function setClient($client)
+    {
         $this->_client = $client;
 
         return $this->_client;
```
Comment (ADmad) [inline src/AbstractDriver.php]: `return $this`. Please just review and make all setter methods return $this :)

**Comment by davidyell** [comment]
Yeah, I've been bust at work with other things. I was going to post to say that I had not forgotten it but thought that 4 days was less than a week, so was probably alright. I'll be pushing up some more commits shortly.

**Comment by ADmad** [comment]
No worries :)

**Comment by davidyell** [comment]
Seems Travis has failed on PHP Stan, and it's a very odd warning that I've been unable to get rid of. I ended up including a `phpstan.neon` with an ignore for it, as even adding a `@var` comment didn't satisfy it.

**Comment by ADmad** [comment]
We can take care of phpstan errors at the end once all changes are done.

**Comment by ADmad** [comment]
Would prefer to have better test coverage for this before merging.

**Comment by davidyell** [comment]
I have noticed a few issues which might be worth adding to  #51 

`Endpoint::getDisplayField` calls `Endpoint::getPrimaryKey()` which calls `Connection::getSchema()`, so the exception will be thrown higher up the stack, so `Endpoint::getDisplayField()` could have it's exception removed.

Many methods in the Endpoint class throw `\Muffin\Webservice\Exception\UnexpectedDriverException` is not as obvious as something like`MissingSchemaException`. Why the Endpoint methods are throwing Driver exceptions I'm unsure of.

File: src/Schema.php [hunk 9]
```diff
@@ -17,7 +17,7 @@ class Schema
     /**
      * The name of the endpoint
      *
-     * @var string
+     * @var \Muffin\Webservice\Model\Endpoint
```
Comment (ADmad) [inline src/Schema.php]: This doesn't seem right. There's `$this->_repository = $endpoint;` inside the class constructor and `$endpoint` is a string, so `$this->_repository` is also a string.

File: src/AbstractDriver.php [hunk 10]
```diff
@@ -271,6 +271,7 @@ public function __debugInfo()
      * @param string $className Class name of the webservice to initialize
      * @param array $options Set of options to pass to the constructor
      * @return WebserviceInterface
+     * @throws MissingWebserviceClassException If no webservice class can be found
```
Comment (ADmad) [inline src/AbstractDriver.php]: FQCN needed.

File: src/Schema.php (hunk 9 above)
Comment (davidyell) [reply src/Schema.php]: Perhaps I've made a mistake in my test case. I will check and push any required updates.

Comment (davidyell) [reply src/Schema.php]: Yes, this is a mistake on my part. This change should have been in my Schema testing branch. I will amend.

**Comment by davidyell** [comment]
This pull request is completed now. Hopefully it's okay to merge. Please let me know if there are any further required changes.

File: src/AbstractDriver.php [hunk 11]
```diff
@@ -194,8 +270,8 @@ public function __debugInfo()
      *
      * @param string $className Class name of the webservice to initialize
      * @param array $options Set of options to pass to the constructor
-     *
      * @return WebserviceInterface
+     * @throws MissingWebserviceClassException If no webservice class can be found
```
Comment (ADmad) [inline src/AbstractDriver.php]: FQCN required.

File: src/Model/Endpoint.php [hunk 12]
```diff
@@ -213,21 +220,43 @@ public function initialize(array $config)
...
+    /**
+     * Set the name of this endpoint
+     *
+     * @param string $name The name for this endpoint instance
+     * @return $this
+     */
+    public function setName($name)
+    {
+        $inflectMethod = $this->getInflectionMethod();
+        $this->_endpoint = Inflector::$inflectMethod($name);
```
Comment (ADmad) [inline src/Model/Endpoint.php]: It would be better to also rename the property `$_endpoint` to `$_name`. To maintain BC using magic `__get()/__set()` to get / set `$_endpoint`.

**Comment by ADmad** [comment]
@davidyell I have fixed the merged conflicts created by change I made to `dev`. You should run `git pull --rebase` locally to update your local `refactor/mixed-mode-methods` branch.

File: tests/TestCase/QueryTest.php [hunk 13]
```diff
@@ -268,7 +268,7 @@ public function testSelectWithExpression()
         $exp = new Comparison('upvotes', 50, 'integer', '>=');
         $this->query->select($exp);
 
-    /** @var Comparison $comparisonClause */
+        /** @var Comparison $comparisonClause */
```
Comment (davidyell) [inline tests/TestCase/QueryTest.php]: Whoops, I think this might have been me accidentally running my global phpcs, rather than the one I have in the project.

Comment (ADmad) [reply tests/TestCase/QueryTest.php]: It's actually correctly indented now :)  We should be using cake's codesniffer v3 if not already using.
